### PR TITLE
Start implementing the `RPNGTemplate` library

### LINKED ISSUE #416: Create a `RPNGTemplate` library
Issue #412 should:

- make the previously widely used `tqec.templates.template.Template` base class an implementation detail,
- expose a new API (called `RPNGTemplate` in #412 but that name might change) that is basically the same as `Template` except that `instantiate` now returns a 2-dimensional array of `RPNG⁴` strings (or `RGN - RPNG⁴`, see #411).

Once #412 will be merged with the new API, we will have to create a few instances of `RPNGTemplate` in order to be able to use them on concrete examples. 

As an example, if the example API in #412 is the final one, below is the implementation of the regular logical qubit
image
performing a memory experiment:

```py
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Literal, Mapping

from tqec.plaquette.frozendefaultdict import FrozenDefaultDict
from tqec.templates.base import Template
from tqec.templates.qubit import QubitTemplate

# Defined in #411 
class BasisEnum(Enum):
    X = "X"
    Y = "Y"
    Z = "Z"

# Defined in #411 
@dataclass
class RPNG:
    r: BasisEnum | None
    p: BasisEnum | None
    n: int | None
    g: BasisEnum | Literal["H"] | None

# Defined in #411
@dataclass
class RPNGPlaquette:
    corners: tuple[RPNG, RPNG, RPNG, RPNG]

    @staticmethod
    def from_string(repr: str) -> RPNGPlaquette: ...

# Defined in #412
@dataclass
class RPNGTemplate:
    template: Template
    mapping: Mapping[int, RPNGPlaquette]

    def instantiate(self, k: int) -> list[list[RPNGPlaquette]]:
        indices = self.template.instantiate(k)
        return [[self.mapping[i] for i in row] for row in indices]


empty_plaquette = RPNGPlaquette.from_string("--- ---- ---- ---- ----")

logical_qubit_memory = RPNGTemplate(
    template=QubitTemplate(),
    mapping=FrozenDefaultDict(
        {
            6: RPNGPlaquette.from_string("xx5 ---- ---- -x3- -x4-"),  # XX_UP
            7: RPNGPlaquette.from_string("xx5 ---- -z3- ---- -z4-"),  # ZZ_LEFT
            9: RPNGPlaquette.from_string("xx5 -x1- -x2- -x3- -x4-"),  # XXXX
            10: RPNGPlaquette.from_string("xx5 -z1- -z3- -z2- -z4-"),  # ZZZZ
            12: RPNGPlaquette.from_string("xx5 -z1- ---- -z2- ----"),  # ZZ_RIGHT
            13: RPNGPlaquette.from_string("xx5 -x1- -x2- ---- ----"),  # XX_DOWN
        },
        default_factory=lambda: empty_plaquette,
    ),
)
```

The above code is only pseudo-code! #411 and #412 might change the actual implementation details, API or class names.

This issue is about using what has been defined in #411 and then in #412 to build usable `RPNGTemplate` instances.

### PR DESCRIPTION
**Pull request by nelimee**
This PR start the implementation of all the `RPNGTemplate` instances we might need to implement a spatial junction as a library. It is based on #421 that should be reviewed before.

Summary of what has been done:

- Removing a check from `RPNG` preventing a plaquette with 3 qubits to be created,
- Modification of `Qubit4WayJunctionTemplate`: renaming as `QubitSpatialJunctionTemplate`, and the implementation is now generic enough to handle all the edge cases (L-shaped spatial junctions, {2,3,4}-way spatial junctions),
- Re-introduction of the "border" templates `QubitVerticalBorder` and `QubitHorizontalBorder`,
- Implementation of a library of `RPNGTemplate`.

In terms of implementation, here are the pieces that are currently implemented:

- [x] Standard memory with optional reset/measurements on data-qubits and both observable orientation (Z observable vertical or horizontal),
- [x] Standard vertical/horizontal junctions, with optional reset/measurements on data-qubits and both observable orientation,
- [x] Temporal hadamard junction,
- [x] Spatial hadamard junction,
- [x] Logical qubit at the center of a spatial junction,
- [ ] Specific spatial junctions to keep plaquettes border with even parity.

If you want to implement any of the missing piece, feel free to 1) say it in a message here and 2) push it directly on this PR :)

Fixes #416 

---------

Notes for future reviews:

- The temporal hadamard is the memory with `R` set to `-` and `M` set to `h`. Is it worth keeping the current implementation, or should I use the memory implementation?
- The spatial hadamard junctions could also be implemented from the memory junctions, but this time it would require to edit the RPNG notations in-place and after creation or to make the memory functions even more complex to allow for a basis change inside plaquettes. Is it worth keeping the current implementation, or should I use the memory implementation?
- I am not fully satisfied with the parameters of the hadamard spatial junctions. For the functions returning the `RPNGTemplate` instances for memory operation, the input is the orientation of the Z observable. That is a very error-prone input in my opinion for hadamard junctions, as their main effect is to change observables... The best I could find was a boolean indicating if the top-left part of the top-left-most plaquette in the bulk (i.e., not on the border) is measuring a Z stabilizer or not. I find that description too complex, but could not find a better way to unambiguously define the template.

### CODE REVIEW

File: src/tqec/templates/indices/qubit.py [hunk 1]
```diff
@@ -70,22 +70,22 @@ def expected_plaquettes_number(self) -> int:
         return 14
 
 
-class Qubit4WayJunctionTemplate(RectangularTemplate):
+class QubitSpatialJunctionTemplate(RectangularTemplate):
```
Comment (inmzhang) [inline src/tqec/templates/indices/qubit.py:73]: Initially it's kind of surprising to me that this template would capture all the spatial junction cases as there is some asymmetry in the template, e.g. `17` is symmetry around `5`(`8`) at the top-left(bottom-right) corner while `13`/`14`(`15`/`16`) is not at the top-right(bottom-left) corner. Then I realized that it's because `5` and `7` are of different basis and the spatial junction only needs to handle hook errors of one basis.

**Comment by inmzhang** [comment]
>     * The temporal hadamard is the memory with `R` set to `-` and `M` set to `h`. Is it worth keeping the current implementation, or should I use the memory implementation?
> 
>     * The spatial hadamard junctions could also be implemented from the memory junctions, but this time it would require to edit the RPNG notations in-place and after creation or to make the memory functions even more complex to allow for a basis change inside plaquettes. Is it worth keeping the current implementation, or should I use the memory implementation?

I did not review the implementations yet. But I will vote for decoupling the implementation of hadamard templates and memory templates.

File: src/tqec/templates/indices/qubit.py (hunk 1 above)
Comment (nelimee) [reply src/tqec/templates/indices/qubit.py:73]: The current implementation might not cover all the cases, but I think it can. The only way to know for sure is by implementing the cases and see if that is possible. This should be tested in the different RPNGTemplate instantiations in the library.

**Comment by nelimee** [comment]
This PR should be ready for review.

File: src/tqec/templates/library/spatial.py [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+    external_stabilizers: Literal["x", "z"],
+    arms: JunctionArms,
+    reset: ResetBasis | None = None,
+    measurement: MeasurementBasis | None = None,
+) -> RPNGTemplate:
+    """Implementation of a logical qubit performing a spatial junction.
+
+    This function returns a RPNGTemplate instance representing a logical qubit
+    that is touched by 2 or more spatial junctions. The returned template is
+    carefully crafted to avoid hook errors damaging the logical distance.
```
Comment (inmzhang) [inline src/tqec/templates/library/spatial.py:29]:
```suggestion
    """Implementation of a logical qubit acting as a spatial junction.

    This function returns a RPNGTemplate instance representing a logical qubit
    connecting to others in space and forming a spatial junction. The returned template is
    carefully crafted to avoid hook errors damaging the logical distance.
```

File: src/tqec/templates/library/spatial.py [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+        By convention, this function does not populate the plaquettes on the
+        boundaries where an arm (i.e. spatial junction) is present **BUT** do
+        populate the corners (that are part of the boundaries, so this is an
+        exception to the first part of the sentence).
+
+        The rationale behind that convention is that the logical qubit at the
+        center of the spatial junction is completely aware of all the spatial
+        junctions that should be implemented whereas each spatial junction in
+        isolation does not know if the others spatial junctions are present or
+        not. That means that corners, whose plaquette depends on the presence or
```
Comment (inmzhang) [inline src/tqec/templates/library/spatial.py:46]: In my view, a spatial junction refers to the logical qubit located at the center of an "X," "T," or "L"-shaped structure. The terminology used here in the documentation is not clear to me.

File: src/tqec/templates/library/spatial.py [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+        shared with the corner, if needed.
+
+        The rationale behind that convention is that the information about
+        whether or not the internal data-qubits of a junction should be reset
+        or measured is given to the junction-generation function, and not to
+        the qubit-generation function (this one). It is also important to note
+        that the information we need here is not given by ``reset`` or
+        ``measurement``: the logical qubit at the center of the spatial junction
+        can already be initialised (and so have an incoming temporal junction
+        from below), which means ``reset is None`` and ``measurement is None``,
```
Comment (inmzhang) [inline src/tqec/templates/library/spatial.py:69]: A spatial junction should not have temporal pipes connecting to them. Otherwise, the boundary conditions will not match.

**Review by inmzhang — COMMENTED** [review]
I haven't finished reviewing the spatial junctions yet, but I will complete it today.

**Comment by afowler** [comment]
On Sun, Dec 22, 2024, 8:49 PM Yiming Zhang ***@***.***> wrote:

> ***@***.**** commented on this pull request.
>
> I haven't finished reviewing the spatial junctions yet, but I will
> complete it today.
> ------------------------------
>
> On src/tqec/templates/rpng_test.py
> <https://github.com/tqec/tqec/pull/422#discussion_r1895179225>:
>
> Empty file
> ------------------------------
>
> In src/tqec/templates/library/memory.py
> <https://github.com/tqec/tqec/pull/422#discussion_r1895184387>:
>
> > +    """Implementation of standard memory rounds on a junction aligned with the
> +    X-axis.
>
> ⬇️ Suggested change
>
> -    """Implementation of standard memory rounds on a junction aligned with the
> -    X-axis.
> +    """Implementation of standard memory rounds on a junction arm aligned with the
> +    X-axis.
>
> ------------------------------
>
> In src/tqec/templates/library/memory.py
> <https://github.com/tqec/tqec/pull/422#discussion_r1895184500>:
>
> > +    """Implementation of standard memory rounds on a junction aligned with the
> +    Y-axis.
>
> ⬇️ Suggested change
>
> -    """Implementation of standard memory rounds on a junction aligned with the
> -    Y-axis.
> +    """Implementation of standard memory rounds on a junction arm aligned with the
> +    Y-axis.
>
> ------------------------------
>
> In src/tqec/templates/library/spatial.py
> <https://github.com/tqec/tqec/pull/422#discussion_r1895229219>:
>
> > +    """Implementation of a logical qubit performing a spatial junction.
> +
> +    This function returns a RPNGTemplate instance representing a logical qubit
> +    that is touched by 2 or more spatial junctions. The returned template is
> +    carefully crafted to avoid hook errors damaging the logical distance.
>
> ⬇️ Suggested change
>
> -    """Implementation of a logical qubit performing a spatial junction.
> -
> -    This function returns a RPNGTemplate instance representing a logical qubit
> -    that is touched by 2 or more spatial junctions. The returned template is
> -    carefully crafted to avoid hook errors damaging the logical distance.
> +    """Implementation of a logical qubit acting as a spatial junction.
> +
> +    This function returns a RPNGTemplate instance representing a logical qubit
> +    connecting to others in space and forming a spatial junction. The returned template is
> +    carefully crafted to avoid hook errors damaging the logical distance.
>
> ------------------------------
>
> In src/tqec/templates/library/spatial.py
> <https://github.com/tqec/tqec/pull/422#discussion_r1895235257>:
>
> > +        The rationale behind that convention is that the logical qubit at the
> +        center of the spatial junction is completely aware of all the spatial
> +        junctions that should be implemented whereas each spatial junction in
> +        isolation does not know if the others spatial junctions are present or
> +        not. That means that corners, whose plaquette depends on the presence or
>
> In my view, a spatial junction refers to the logical qubit located at the
> center of an "X," "T," or "L"-shaped structure. The terminology used here
> in the documentation is not clear to me.
>
+1, agreed

------------------------------
>
> In src/tqec/templates/library/spatial.py
> <https://github.com/tqec/tqec/pull/422#discussion_r1895239034>:
>
> > +        ``measurement``: the logical qubit at the center of the spatial junction
> +        can already be initialised (and so have an incoming temporal junction
> +        from below), which means ``reset is None`` and ``measurement is None``,
>
> A spatial junction should not have temporal pipes connecting to them.
> Otherwise, the boundary conditions will not match.
>
+1, you can't mix spatial and temporal junctions. It's always either one or
the other.

File: src/tqec/templates/library/spatial.py (hunk 3 above)
Comment (nelimee) [reply src/tqec/templates/library/spatial.py]: I agree. The terminology I will use from now on is "spatial junction" for the logical qubit performing the junction and "arm" for the pipes around the spatial junction. Does that seems reasonable to you?

File: src/tqec/templates/library/spatial.py (hunk 4 above)
Comment (nelimee) [reply src/tqec/templates/library/spatial.py]: I just removed that assumption on the spatial junction (i.e., the logical qubit). It might also simplify the arms implementation, but I have not done it yet. If you see a potential improvement in your review, please let me know.

File: src/tqec/templates/library/spatial.py [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+    return RPNGTemplate(
+        template=QubitSpatialJunctionTemplate(),
+        mapping=FrozenDefaultDict(
+            mapping,
+            default_factory=lambda: RPNGDescription.from_string("---- ---- ---- ----"),
+        ),
+    )
+
+
+def get_spatial_junction_junction_template(
```
Comment (inmzhang) [inline src/tqec/templates/library/spatial.py:221]:
```suggestion
def get_spatial_junction_arm_template(
```
Some terminology in the documentation also needs to be revised.

File: src/tqec/templates/library/spatial.py [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+    carefully crafted to avoid hook errors damaging the logical distance.
+
+    Note:
+        this function does not enforce anything on the input values. As such, it
+        is possible to generate a description of a round that will both reset and
+        measure the data-qubits.
+
+    Warning:
+        By convention, this function does not populate the plaquettes on the
+        boundaries where an arm (i.e. spatial junction) is present **BUT** do
```
Comment (inmzhang) [inline src/tqec/templates/library/spatial.py:38]:
```suggestion
        boundaries where an arm is present **BUT** do
```

File: src/tqec/templates/library/spatial.py [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+    Arguments:
+        external_stabilizers: stabilizers that are measured at each boundaries
+            of the spatial junction.
+        arm: arm to return a spatial junction for. Should contain exactly
+            **one** of the possible arm flags.
+        reset: basis of the reset operation performed on data-qubits. Defaults
+            to ``None`` that translates to no reset being applied on data-qubits.
+        measurement: basis of the measurement operation performed on data-qubits.
+            Defaults to ``None`` that translates to no measurement being applied
+            on data-qubits.
```
Comment (inmzhang) [inline src/tqec/templates/library/spatial.py:259]:
```suggestion
        reset: basis of the reset operation performed on **internal** data-qubits. Defaults
            to ``None`` that translates to no reset being applied on data-qubits.
        measurement: basis of the measurement operation performed on **internal** data-qubits.
            Defaults to ``None`` that translates to no measurement being applied
            on data-qubits.
```

File: src/tqec/templates/library/spatial_test.py
Comment (inmzhang) [inline src/tqec/templates/library/spatial_test.py:1]: Tests for the `k=1` edge case should be added.

**Review by inmzhang — COMMENTED** [review]
It's quite a lot of work, well done! The implementation looks correct to me, and the tests appear robust. Over the next few days, I plan to implement SVG visualization for the `RPNGTemplate` to make it easier to visually inspect the templates. Additionally, I believe implementing a graph algorithm to find the shortest logical operator across the spatial junction, based on counting (hook) errors, is feasible and not overly complex given the `RPNGTemplate`. I might explore creating such an algorithm and incorporating it into the tests to further verify correctness.

**Comment by nelimee** [comment]
> Additionally, I believe implementing a graph algorithm to find the shortest logical operator across the spatial junction, based on counting (hook) errors, is feasible and not overly complex given the `RPNGTemplate`. I might explore creating such an algorithm and incorporating it into the tests to further verify correctness.

I also though about this when implementing the `RPNGTemplate`. I agree that this would be welcomed, and it might even be a first step toward automatic hook error management.

File: src/tqec/templates/library/spatial.py [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+from tqec.templates.indices.qubit import (
+    QubitHorizontalBorders,
+    QubitSpatialJunctionTemplate,
+    QubitVerticalBorders,
+)
+from tqec.templates.rpng import RPNGTemplate
+
+
+def get_spatial_junction_qubit_template(
+    external_stabilizers: Literal["x", "z"],
```
[the remark this answers is not in the document]
Comment (nelimee) [reply src/tqec/templates/library/spatial.py:20]: I agree that it would also work. I'll change that.

File: src/tqec/templates/library/spatial.py [hunk 9]
```diff
@@ -0,0 +1,430 @@
...
+    # corners (i.e. indices {5, 6, 7, 8}) should be assigned "regular" plaquettes
+    # (i.e. 6 and 7 have the same plaquette as 17, 5 has the same plaquette as
+    # 13 and 8 has the same plaquette as 15). If these need to be changed, it
+    # will be done afterwards.
+    internal_basis_plaquette = RPNGDescription.from_string(
+        f"{r}{bi}1{m} {r}{bi}3{m} {r}{bi}2{m} {r}{bi}5{m}"
+    )
+    mapping[6] = internal_basis_plaquette
+    mapping[7] = internal_basis_plaquette
+    mapping[17] = internal_basis_plaquette
```
[the remark this answers is not in the document]
Comment (nelimee) [reply src/tqec/templates/library/spatial.py:173]: At first, it was for future-proofing. I wanted an expressive enough template to implement everything, and made it a little bit too expressive. Now, it is really for symmetry of the template (which has no value at all implementation-wise).

It can be changed, I have no preference on that.

**Review by inmzhang — APPROVED** [review]
